### DISCUSSION TITLE
TypeError: before_cursor_execute() missing 1 required positional argument: 'executemany'

### DISCUSSION THREAD

**Discussion by munaAchyuta** [sqlalchemy/sqlalchemy#7509] [Usage Questions] [answered]
### Describe the bug

i'm getting this issue(TypeError: before_cursor_execute() missing 1 required positional argument: 'executemany') recently. i had created separate conda environment and installed required packages for my project. but then i created a new conda environment having installed same package versions as with other working one. but this time sqlalchemy complain with above error. 

even i tried alternative approaches like uninstall and install of anaconda and then fresh start of environment, still i get same issue. also i tried using python base , but still get same issue.

it was working fine. with old environment. but now it's giving error even when i tried other versions(2.*) of sqlalchemy. 

how to resolve this issue ? 

### To Reproduce

```python
import uuid
from datetime import datetime
from sqlalchemy import (
    ARRAY,
    Boolean,
    Column,
    DateTime,
    ForeignKey,
    Integer,
    MetaData,
    String,
    Text,
    JSON,
    CheckConstraint,
    Float,
    ForeignKeyConstraint
)

from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import relationship
from sqlalchemy.schema import UniqueConstraint
import logging
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker

class Audit:
    createdDate = Column("created_date", DateTime, default=datetime.now())
    updatedDate = Column(
        "updated_date",
        DateTime,
        default=datetime.now(),
        onupdate=datetime.now(),
    )
    createdBy = Column("created_by", String, default="damon")
    updatedBy = Column("updated_by", String, default="damon")

#...and many other class models defined.

url = "postgresql+psycopg2://{user}:{passwd}{url}".format(user="xyz",passwd="xyz",url="@host:port/db_name",)
engine = create_engine(url, pool_size=50, echo=False)

Base.metadata.create_all(engine, checkfirst=True)
```

### Error

```
# Copy complete stack trace and error message here, including SQL log output if applicable.
File "C:\Users\anama\Anaconda3\envs\ms_env\lib\site-packages\sqlalchemy\sql\ddl.py", line 788, in _can_create_table
return not self.checkfirst or not self.dialect.has_table(
File "C:\Users\anama\Anaconda3\envs\ms_env\lib\site-packages\sqlalchemy\dialects\postgresql\base.py", line 3357, in has_table
cursor = connection.execute(
File "C:\Users\anama\Anaconda3\envs\ms_env\lib\site-packages\sqlalchemy\engine\base.py", line 1286, in execute
return meth(self, multiparams, params, _EMPTY_EXECUTION_OPTS)
File "C:\Users\anama\Anaconda3\envs\ms_env\lib\site-packages\sqlalchemy\sql\elements.py", line 323, in _execute_on_connection
return connection._execute_clauseelement(
File "C:\Users\anama\Anaconda3\envs\ms_env\lib\site-packages\sqlalchemy\engine\base.py", line 1478, in _execute_clauseelement
ret = self._execute_context(
File "C:\Users\anama\Anaconda3\envs\ms_env\lib\site-packages\sqlalchemy\engine\base.py", line 1741, in _execute_context
statement, parameters = fn(
File "C:\Users\anama\Anaconda3\envs\ms_env\lib\site-packages\sqlalchemy\engine\events.py", line 128, in wrap_before_cursor_execute
orig_fn(
TypeError: before_cursor_execute() missing 1 required positional argument: 'executemany'
``` 

### Versions

- OS: Linux
- Python: 3.7
- SQLAlchemy: 1.4.25 & 1.4.29
- Database: Postgres
- DBAPI (eg: psycopg, cx_oracle, mysqlclient): psycopg2

**Comment by zzzeek** [comment, marked as answer]
hey there -

the code given seems to be a paraphrase of part of your application as it isn't complete.    In any case, SQLAlchemy doesn't use the "before_cursor_execute" event internally, so you would need to locate the "before_cursor_execute" listener or listeners in your program and make sure it has [the correct signature](https://docs.sqlalchemy.org/en/14/core/events.html?highlight=before_cursor_execute#sqlalchemy.events.ConnectionEvents.before_cursor_execute) including the "executemany" parameter.

**Reply by munaAchyuta** [reply]
Hi @zzzeek ,

i know i have shared paraphrase of part of my application. but this is only i use to build my SQL ORM model and call create_all() method to check if tables already exists or not.

if sqlalchemy doesn't use before_cursor_execute event internally, then why does that issue coming ? 
here is where issue coming...

`File "C:\Users\anama\Anaconda3\envs\ms_env\lib\site-packages\sqlalchemy\engine\events.py", line 128, in wrap_before_cursor_execute
orig_fn(
TypeError: before_cursor_execute() missing 1 required positional argument: 'executemany'`

and why do i need to add before_cursor_execute listner in my program ?  i'm not using/calling this event anywhere in my application.

any other functionality causing this issue ?

**Reply by zzzeek** [reply]
> Hi @zzzeek ,
> 
> i know i have shared paraphrase of part of my application. but this is only i use to build my SQL ORM model and call create_all() method to check if tables already exists or not.
> 
> if sqlalchemy doesn't use before_cursor_execute event internally, then why does that issue coming ?

i dont know, but it would appear something you have installed in your environment is using this event handler.  it would not be invoked with the code you have above. 


> here is where issue coming...
> 
> `File "C:\Users\anama\Anaconda3\envs\ms_env\lib\site-packages\sqlalchemy\engine\events.py", line 128, in wrap_before_cursor_execute orig_fn( TypeError: before_cursor_execute() missing 1 required positional argument: 'executemany'`

specifically it looks like a before_cursor_execute() event handler funciton that has one too many arguments in its signature.


> 
> and why do i need to add before_cursor_execute listner in my program ? i'm not using/calling this event anywhere in my application.

something in your environment is using it.    

I suggest you use pdb to inspect the function.   Here's an example program illustrating how to get the error:

```py
from sqlalchemy import create_engine, event, text


e = create_engine("sqlite://")


@event.listens_for(e, "before_cursor_execute")
def before_cursor_execute_one_arg_too_many(
    too_many, conn, cursor, statement, parameters, context, executemany
):
    pass


with e.connect() as conn:
    try:
        conn.execute(text("select 1"))
    except:
        import pdb
        pdb.post_mortem()
```

the important part is to identify the line in your program where the stack trace for the above error begins.   If we run the program above, we get a "pdb" prompt.  orig_fn, the funciton in question, will be right there, and you can inspect `__code__` which will give you the file and line number where the function is present:

```
$ python test3.py 
> /home/classic/.venv3/lib/python3.10/site-packages/sqlalchemy/engine/events.py(128)wrap_before_cursor_execute()
-> orig_fn(
(Pdb) locals()
{'conn': <sqlalchemy.engine.base.Connection object at 0x7fc713475c90>, 'cursor': <sqlite3.Cursor object at 0x7fc711dec040>, 'statement': 'select 1', 'parameters': (), 'context': <sqlalchemy.dialects.sqlite.base.SQLiteExecutionContext object at 0x7fc711dbd3f0>, 'executemany': False, 'orig_fn': <function before_cursor_execute_one_arg_too_many at 0x7fc711de5360>}
(Pdb) orig_fn
<function before_cursor_execute_one_arg_too_many at 0x7fc711de5360>
(Pdb) orig_fn.__code__
<code object before_cursor_execute_one_arg_too_many at 0x7fc713523470, file "/home/classic/dev/sqlalchemy/test3.py", line 7>
(Pdb) 

```

**Reply by CaselIT** [reply]
@zzzeek I guess we could try surfacing the __code__ information in case of type errors in invoking an event?
Or do you thinking the performance impact would be too large?

**Reply by zzzeek** [reply]
well you mean an exception handler, so there'd be no performance impact.

i just dont know that an open ended exception handler isn't more often going to get in the way of reporting normal runtime problems in event handlers.    it's very unusual for an environment that has a non-working event handler where it isn't fairly obvious where that handler is.

**Reply by CaselIT** [reply]
> i just dont know that an open ended exception handler isn't more often going to get in the way of reporting normal runtime problems in event handlers.

yes, that's probably true. I guess we could limit it to `TypeError` but even that it's quite a broad case, since type errors are quite common
